juju/terraform-provider-juju
# Add resource block to juju_application

## Description

fix #194 
This change adds the block `resource` to the resource `juju_application`, this block is repeatable and only support setting oci-images.

Example:

```hcl
resource "juju_application" "ex" {
  ...
  resource {
    name      = "charm-oci-image"
    oci_image = "registry:5000/my_app/my_image"
  }
}
````

## Type of change

Please mark if proceeds.


## Environment

- Juju controller version: 2.9.42
- Terraform version: v1.4.5

## QA steps

Manual QA steps should be done to test this PR.

- Step 1 `juju add-model resource`
- Step 2 run terraform apply with the following snippet, check the payload container has the correct image
```tf
provider juju {}

data "juju_model" "resource" {
  name = "resource"
}

resource "juju_application" "resource_example" {
  name  = "resource-example"
  model = data.juju_model.resource.name
  charm {
    name     = "hello-kubecon"
    channel  = "edge"
    revision = 19
    series   = "focal"
  }

  resource {
    name = "gosherve-image"
    oci_image = "jnsgruk/gosherve:latest"
  }
}
```
- Step 3: Delete the inner resource block and apply, note the application is deleted and recreated with the charmstore resource

# Additional notes
- There's no infrastructure for testing K8S charms, and this would require specific K8S tests
- Change in behavior of DeleteResource, because force new would try to re-create the resource as soon as the delete call was sent to the controller. The new create request would fail because the application is still present
